### PR TITLE
feat: upscale with the Stable Diffusion x4 upscaler model

### PR DESCRIPTION
**Pull request by keturn**
This adds a node to upscale using the [Stable Diffusion x4 upscaler model](https://huggingface.co/stabilityai/stable-diffusion-x4-upscaler).

Personally, I don't know whether it's better than ESRGAN, but I know some people have wanted to try it out.

**Not to be confused with [Stable Diffusion x2 latent upscaler](https://huggingface.co/stabilityai/sd-x2-latent-upscaler).** The x2 one can take normal SD latents and upscale them. This x4 one has its own VAE, so it's not interchangeable with SD latents. For this reason, I've implemented it as a single image→image node instead of a series of encode→diffusion→decode nodes.

Currently in proof-of-concept stage. Some open considerations:

- Code duplication.
  + [StableDiffusionUpscalePipeline](https://github.com/huggingface/diffusers/blob/5e80827369272f4b24af51573466aba24454b068/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py#L68) is _almost_ the same as the normal Stable Diffusion pipeline, so we might be able to re-use more of our normal SD pipeline code instead of these upscaler-specific invocations. But like most diffusers pipeline code, it is very hard to tell which of those 800 lines are actually unique to that pipeline instead of copies of stuff already implemented in other pipelines.
  + If we do need this new Invocation class, there might be some chunks to factor out relating to the VAE data type, and also determining `source_node_id`.
- Reproducible noise. Should expose the model's `noise_level` parameter and wire up the `noise` input.
- Give the option to disable guidance? We _can_ pass text embeddings to this thing for it to use for the familiar classifier-free guidance method, but since an upscaler is limited in the scope of the changes it can make, it's possible that it's not worth it. We could set guidance scale to 1.0 and find out if the cost of doing guidance is worthwhile. (Could be up to twice as fast without it.)
- Progress preview is borked, since it's a different VAE. Not that the preview is super useful, because low-res previews don't give much information on how the high-res enhancement looks — but it's nice to have a window into the progress in any case.
- The upscaler shows up in the UI's model list as a "Stable Diffusion 2.x" model, but trying to use it as one results only in an error.

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [x] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [ ] Yes
- [x] No, because: discussion starts here! See also [accompanying share-your-nodes post](https://discord.com/channels/1020123559063990373/1133588159616270387/1133588159616270387).

      
## Have you updated all relevant documentation?
- [ ] Yes
- [x] No

## Related Tickets & Documents

- Related Issue #2254

## QA Instructions, Screenshots, Recordings

- use model manager to import `stabilityai/stable-diffusion-x4-upscaler`
- use "Upscale (Stable Diffusion x4)" node to turn an image in to a bigger image

## Added/updated tests?

- [ ] Yes
- [x] No : not yet!

### CODE REVIEW

**Comment by keturn** [comment]
Example node graph: https://gist.github.com/keturn/898759e06370a4786974d766d1835eef#file-sd-x4-upscaler-invokeai-json

**Comment by keturn** [comment]
<details>
<summary>sample output:</summary>

[20 steps, DPMSolver++ k]

sdx4-output

compare to input:

upscaling-in

output of some previous method: 

upscaling-out

</details>

**Comment by lstein** [comment]
We probably need to add a new "upscaler" model type so that this kind of model doesn't contaminate the main model list. I see that it has a distinct class name of `StableDiffusionUpscalePipeline`, so it will be easy to add this to the model type probe module. I'll do that in the next couple days.

**Comment by lstein** [comment]
Is this a core model that should be added at install time, or an optional model that people can choose to install? I suspect the latter.

**Comment by keturn** [comment]
Depends on how we make that decision about upscalers?

On the one hand, the ESRGAN models are always installed despite the fact that I never use them.

On the other hand, this SD x4 upscaler is 3.3 GB — 16x more disk space than all the ESRGAN models _combined._ I'm certainly okay with that being optional.

File: invokeai/app/invocations/upscale_sdx4.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    def invoke(self, context: InvocationContext) -> ImageOutput:
+        image = context.services.images.get_pil_image(self.image.image_name)
+
+        # TODO: are there upscaler LoRA? ControlNet? document if not
+
+        model_manager = context.services.model_manager
+        unet_info = model_manager.get_model(**self.unet.unet.dict(), context=context)
+        vae_info = model_manager.get_model(**self.vae.vae.dict(), context=context)
+
+        # ¿what is this context manager doing?
```
Comment (psychedelicious) [inline invokeai/app/invocations/upscale_sdx4.py:56]: the context manages the loading of the model into and out of VRAM

**Review by psychedelicious — COMMENTED** [review]
happy to see you back in invoke-land!

File: invokeai/app/invocations/upscale_sdx4.py (hunk 1 above)
Comment (keturn) [reply invokeai/app/invocations/upscale_sdx4.py:56]: oh, I'd looked up the wrong `ModelInfo` class. Now I see the one with the [context manager methods](https://github.com/invoke-ai/InvokeAI/blob/be00a837cce6072ea7bebc93ba4e8ac1cfa81fac/invokeai/backend/model_management/model_manager.py#L275). That makes sense, thanks!

**Comment by hipsterusername** [comment]
@lstein @RyanJDick - This has been open for a while. Do we think it should get a pre-MM rewrite update for models? Sounds like that may be further out.

**Comment by RyanJDick** [comment]
> @lstein @RyanJDick - This has been open for a while. Do we think it should get a pre-MM rewrite update for models? Sounds like that may be further out.

From my perspective, the model management integration for this PR is still TODO, so it's not like we would be throwing away work if we waited until after the MM refactor. How much demand is there for this feature? (relative to e.g. T2I Adapters, ControlLoRAs, etc.) It does seem like it would be silly to merge this and then immediately refactor it for the new MM.

(I may be biased, as my recent experience with MM for IP-Adapter makes me really want to prioritize getting the MM refactor in ASAP. 🙃)

**Comment by hipsterusername** [comment]
Frankly, I think Upscaling is a pretty high demand feature, however until we support _tiled_ upscaling I think we'll find that we're still seen as 'subpar' to A1111, so it'll just be closing the gap there a bit.

**Comment by keturn** [comment]
It doesn't look like the upstream pipeline gives us tiling for free, but I don't think it would be one of the more difficult things we've done. Since upscalers focus so much on local detail, I'd expect the required seam overlap to be pretty small.

**Pull request closed without merging**